### PR TITLE
operator: the operator timeout duation depends on all the step not separated

### LINKED ISSUE #5596: The timeout should consider the transfer latency especially in big cluster
## Bug Report
In big cluster, the duration of region heartbeat is bigger than 10s. For example:
the region heartbeat qps for every store : 1w ops
there are 10w region in this store, so the latency is 10w/1w=10s
but some short operator step timeout limit is 10s, such as promote/demote/transfer leader, so the operator failed because the duration is bigger than 10s. 

some log:
```
[2022/10/13 00:52:23.560 +00:00] [INFO] [operator_controller.go:590] ["operator timeout"] [region-id=1820635949] [takes=1m20.047056718s] [operator="\"balance-region {mv peer: store [14057] to [14053]} (kind:region, region:1820635949(540, 1763), createAt:2022-10-13 00:51:03.513592708 +0000 UTC m=+461704.590481929, startAt:2022-10-13 00:51:03.513683079 +0000 UTC m=+461704.590572320, currentStep:1, size:16, steps:[add learner peer 1832839605 on store 14053, use joint consensus, promote learner peer 1832839605 on store 14053 to voter, demote voter peer 1820635950 on store 14057 to learner, leave joint state, promote learner peer 1832839605 on store 14053 to voter, demote voter peer 1820635950 on store 14057 to learner, remove peer on store 14057]) timeout\""] [additional-info="{\"sourceScore\":\"5538250.44\",\"targetScore\":\"5519343.11\"}"]
```
there are 5w regions in each store.

the operator step duration
image
the duration of region heartbeat 
image

### What did you expect to see?
operators are successful.
### What did you see instead?
operators are failed
### What version of PD are you using (`pd-server -V`)?
master, v6.1.0

### PR DESCRIPTION
**Pull request by bufferflies**
Signed-off-by: bufferflies <1045931706@qq.com>

### What problem does this PR solve?
Relax the timeout duration limit for slow operator. 
If the duration of region heartbeat request is higher than 60s , the cluster can not be healthy. 

Issue Number: ref #5596

### What is changed and how does it work?
change the interface `timeout(start time.time,regionSize uint64)bool` to `timeout(regionSize)time.duration`.
1. calculate the max duation depends on the steps 
2. checkout timeout will only check the total duation not all the step

```commit-message
```

### Check List

Tests

- Unit test
- Integration test

Code changes

Side effects

Related changes

- Need to cherry-pick to the release branch

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/grpc_service.go [hunk 1]
```diff
@@ -895,6 +895,7 @@ func (s *GrpcServer) RegionHeartbeat(stream pdpb.PD_RegionHeartbeatServer) error
 			continue
 		}
 
+		transportDuration := time.Now().Unix() - int64(request.GetInterval().GetEndTimestamp())
```
Comment (nolouch) [inline server/grpc_service.go:898]: Do not relay on this, it's may be different system time in the host.

Comment (bufferflies) [reply server/grpc_service.go]: yes, the latency metrics already exits , I will remove it .

**Review by lhy1024 — APPROVED** [review]

**Comment by bufferflies** [comment]
PTAL @nolouch again

File: server/schedule/operator/step.go [hunk 2]
```diff
@@ -112,8 +120,8 @@ func (tl TransferLeader) Influence(opInfluence OpInfluence, region *core.RegionI
 }
 
 // Timeout returns true if the step is timeout.
-func (tl TransferLeader) Timeout(start time.Time, regionSize int64) bool {
-	return time.Since(start) > fastStepWaitDuration(regionSize)
+func (tl TransferLeader) Timeout(regionSize int64) time.Duration {
```
Comment (rleungx) [inline server/schedule/operator/step.go:123]: The comment needs to be changed.

File: server/schedule/operator/step.go [hunk 3]
```diff
@@ -561,8 +569,9 @@ func (mr MergeRegion) Influence(opInfluence OpInfluence, region *core.RegionInfo
 }
 
 // Timeout returns true if the step is timeout.
-func (mr MergeRegion) Timeout(start time.Time, regionSize int64) bool {
-	return time.Since(start) > fastStepWaitDuration(regionSize)*10
+// The merge operator must wait for the first operator finished, so the executing duration must larger than add learner.
+func (mr MergeRegion) Timeout(regionSize int64) time.Duration {
+	return slowStepWaitDuration(regionSize) * 2
```
Comment (rleungx) [inline server/schedule/operator/step.go:574]: Is the change expected?

File: server/schedule/operator/step.go [hunk 4]
```diff
@@ -38,6 +38,14 @@ const (
 	// DefaultFastExecutorRate is the slow rate of the operator executor.
 	// default:  0.1 s/Mb
 	DefaultFastExecutorRate = 0.1
+	// FastOperatorWaitTime is the duration that when an operator that is not marked
```
Comment (rleungx) [inline server/schedule/operator/step.go:41]: Should be changed to XXXStepWaitTime? Also, the comment needs to be changed.

File: server/schedule/operator/step.go (hunk 3 above)
Comment (bufferflies) [reply server/schedule/operator/step.go:574]: yes, the merge step must wait the first operator (move) finished .

File: server/schedule/operator/step.go (hunk 4 above)
Comment (bufferflies) [reply server/schedule/operator/step.go]: fixed

File: server/schedule/operator/step.go (hunk 3 above)
Comment (CabinfeverB) [reply server/schedule/operator/step.go:574]: I feel that * 2 is not enough. The first operator contains multiple move peer steps

File: server/schedule/operator/step.go (hunk 2 above)
Comment (bufferflies) [reply server/schedule/operator/step.go:123]: fixed

File: server/schedule/operator/status_tracker.go [hunk 5]
```diff
@@ -116,11 +116,12 @@ func (trk *OpStatusTracker) CheckExpired(exp time.Duration) bool {
 }
 
 // CheckStepTimeout checks if timeout, and update the current status.
-func (trk *OpStatusTracker) CheckStepTimeout(start time.Time, step OpStep, approximateSize int64) bool {
+func (trk *OpStatusTracker) CheckStepTimeout(duration time.Duration) bool {
```
Comment (rleungx) [inline server/schedule/operator/status_tracker.go:119]:
```suggestion
func (trk *OpStatusTracker) CheckTimeout(duration time.Duration) bool {
```

File: server/schedule/operator/operator.go [hunk 6]
```diff
@@ -56,6 +50,7 @@ type Operator struct {
 	FinishedCounters []prometheus.Counter
 	AdditionalInfos  map[string]string
 	ApproximateSize  int64
+	maxDuration      time.Duration
```
Comment (rleungx) [inline server/schedule/operator/operator.go:53]: How about just timeout?

**Review by rleungx — COMMENTED** [review]
The rest LGTM.

File: server/schedule/operator/status_tracker.go (hunk 5 above)
Comment (bufferflies) [reply server/schedule/operator/status_tracker.go]: fixed

File: server/schedule/operator/operator.go (hunk 6 above)
Comment (bufferflies) [reply server/schedule/operator/operator.go]: fixed

File: server/schedule/operator/operator.go [hunk 7]
```diff
@@ -76,9 +75,15 @@ func NewOperator(desc, brief string, regionID uint64, regionEpoch *metapb.Region
 		level:           level,
 		AdditionalInfos: make(map[string]string),
 		ApproximateSize: approximateSize,
+		timeout:         time.Duration(maxDuration) * time.Second,
 	}
 }
 
+// Sync some attribute with the given timeout.
```
Comment (CabinfeverB) [inline server/schedule/operator/operator.go:82]: Maybe we can use option to specify which fields to sync

File: server/schedule/operator/step.go [hunk 8]
```diff
@@ -38,6 +38,14 @@ const (
 	// DefaultFastExecutorRate is the slow rate of the operator executor.
 	// default:  0.1 s/Mb
 	DefaultFastExecutorRate = 0.1
```
Comment (CabinfeverB) [inline server/schedule/operator/step.go:40]: Should change DefaultFastExecutorRate？because SlowStepWaitTime/FastStepWaitTime is changed to 10 from 60

**Review by CabinfeverB — COMMENTED** [review]
rest LGTM

File: server/schedule/operator/operator.go (hunk 7 above)
Comment (bufferflies) [reply server/schedule/operator/operator.go:82]: In past, only timeout needs to sync. It should be done if sync one more properties

File: server/schedule/operator/step.go (hunk 8 above)
Comment (bufferflies) [reply server/schedule/operator/step.go]: fixed to 0.6

File: server/schedule/operator/operator.go [hunk 9]
```diff
@@ -76,9 +75,15 @@ func NewOperator(desc, brief string, regionID uint64, regionEpoch *metapb.Region
 		level:           level,
 		AdditionalInfos: make(map[string]string),
 		ApproximateSize: approximateSize,
+		timeout:         time.Duration(maxDuration) * time.Second,
```
Comment (CabinfeverB) [inline server/schedule/operator/operator.go:78]: Can we put the timeout into `brief` or `additional-info`

**Review by CabinfeverB — APPROVED** [review]

**Pull request merged**